File: src/JavaScript/devpossible-ton-samples/examples/file-operations.js
/**
 * File Operations Example
 * Copyright (c) 2024 DevPossible, LLC
 */

import * as ton from '../../devpossible-ton/dist/index.esm.js';
import fs from 'fs';
import path from 'path';

export async function run() {
  console.log('=== File Operations Example ===\n');

  // Example 1: Read TON file
  console.log('1. Reading TON file:');
  const simplePath = path.join(process.cwd(), 'SampleData', 'simple.ton');
  
  try {
    const content = fs.readFileSync(simplePath, 'utf8');
    const parser = new ton.TonParser();
    const document = parser.parse(content);
    
    console.log('   File read successfully!');
    console.log('   Content:', JSON.stringify(document.toJSON(), null, 2));
  } catch (error) {
    console.log('   Error reading file:', error.message);
  }

  // Example 2: Write TON file
  console.log('\n2. Writing TON file:');
  const outputPath = path.join(process.cwd(), 'SampleData', 'output.ton');
  
  const data = {
    application: 'Sample App',
    version: '2.0.0',
    settings: {
      theme: 'dark',
      language: 'en',
      notifications: true
    },
    features: ['analytics', 'reporting', 'export']
  };

  const doc = new ton.TonDocument();
  doc.setRoot(data);
  
  const serializer = new ton.TonSerializer();
  const tonContent = serializer.serialize(doc);
  
  try {
    fs.writeFileSync(outputPath, tonContent, 'utf8');
    console.log('   File written successfully to:', outputPath);
  } catch (error) {
    console.log('   Error writing file:', error.message);
  }

  // Example 3: Read and modify
  console.log('\n3. Read, modify, and save:');
  const configPath = path.join(process.cwd(), 'SampleData', 'config.ton');
  
  try {
    const configContent = fs.readFileSync(configPath, 'utf8');
    const parser = new ton.TonParser();
    const configDoc = parser.parse(configContent);
    
    console.log('   Original config loaded');
    
    // Modify the document
    const root = configDoc.getRoot();
    root.set('modified', new Date().toISOString());
    root.set('version', '1.0.1');
    
    // Save modified version
    const modifiedPath = path.join(process.cwd(), 'SampleData', 'config-modified.ton');
    const modifiedContent = serializer.serialize(configDoc);
    fs.writeFileSync(modifiedPath, modifiedContent, 'utf8');
    
    console.log('   Modified config saved to:', modifiedPath);
  } catch (error) {
    console.log('   Error:', error.message);
  }

  // Example 4: Batch processing
  console.log('\n4. Batch file processing:');
  const sampleDir = path.join(process.cwd(), 'SampleData');
  
  try {
    const files = fs.readdirSync(sampleDir)
      .filter(f => f.endsWith('.ton'));
    
    console.log(`   Found ${files.length} TON files:`);
    
    const parser = new ton.TonParser();
    files.forEach(file => {
      try {
        const filePath = path.join(sampleDir, file);
        const content = fs.readFileSync(filePath, 'utf8');
        const document = parser.parse(content);
        console.log(`     ✓ ${file} - Valid TON file`);
      } catch (error) {
        console.log(`     ✗ ${file} - Parse error: ${error.message}`);
      }
    });
  } catch (error) {
    console.log('   Error reading directory:', error.message);
  }
}

File: src/Python/devpossible_ton_samples/examples/file_operations.py
#!/usr/bin/env python
"""
File Operations Example
Copyright (c) 2024 DevPossible, LLC
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'devpossible_ton'))

import devpossible_ton as ton
from datetime import datetime


def run():
    """Run file operations examples."""
    print('=== File Operations Example ===\n')

    # Example 1: Read TON file
    print('1. Reading TON file:')
    simple_path = os.path.join(os.path.dirname(__file__), '..', 'SampleData', 'simple.ton')
    
    try:
        with open(simple_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        parser = ton.TonParser()
        document = parser.parse(content)
        
        print('   File read successfully!')
        print(f'   Content: {document.to_json()}')
    except Exception as error:
        print(f'   Error reading file: {error}')

    # Example 2: Write TON file
    print('\n2. Writing TON file:')
    output_path = os.path.join(os.path.dirname(__file__), '..', 'SampleData', 'output.ton')
    
    data = {
        'application': 'Sample App',
        'version': '2.0.0',
        'settings': {
            'theme': 'dark',
            'language': 'en',
            'notifications': True
        },
        'features': ['analytics', 'reporting', 'export']
    }

    doc = ton.TonDocument()
    doc.set_root(data)
    
    serializer = ton.TonSerializer()
    ton_content = serializer.serialize(doc)
    
    try:
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(ton_content)
        print(f'   File written successfully to: {output_path}')
    except Exception as error:
        print(f'   Error writing file: {error}')

    # Example 3: Read and modify
    print('\n3. Read, modify, and save:')
    config_path = os.path.join(os.path.dirname(__file__), '..', 'SampleData', 'config.ton')
    
    try:
        with open(config_path, 'r', encoding='utf-8') as f:
            config_content = f.read()
        
        parser = ton.TonParser()
        config_doc = parser.parse(config_content)
        
        print('   Original config loaded')
        
        # Modify the document
        root = config_doc.get_root()
        from devpossible_ton.models.ton_object import TonObject
        if isinstance(root, TonObject):
            root.set('modified', datetime.now().isoformat())
            root.set('version', '1.0.1')
        
        # Save modified version
        modified_path = os.path.join(os.path.dirname(__file__), '..', 'SampleData', 'config-modified.ton')
        modified_content = serializer.serialize(config_doc)
        
        with open(modified_path, 'w', encoding='utf-8') as f:
            f.write(modified_content)
        
        print(f'   Modified config saved to: {modified_path}')
    except Exception as error:
        print(f'   Error: {error}')

    # Example 4: Batch processing
    print('\n4. Batch file processing:')
    sample_dir = os.path.join(os.path.dirname(__file__), '..', 'SampleData')
    
    try:
        files = [f for f in os.listdir(sample_dir) if f.endswith('.ton')]
        
        print(f'   Found {len(files)} TON files:')
        
        parser = ton.TonParser()
        for file in files:
            try:
                file_path = os.path.join(sample_dir, file)
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                document = parser.parse(content)
                print(f'     ✓ {file} - Valid TON file')
            except Exception as error:
                print(f'     ✗ {file} - Parse error: {error}')
    except Exception as error:
        print(f'   Error reading directory: {error}')
